### DISCUSSION TITLE
default argument

### DISCUSSION THREAD

**Discussion by hbadi** [pybind/pybind11#3360] [Q&A]
Hi,

Default arguments are not so easy to use in classes with constructor and overloaded methods.

I have the following very basic class

class A {
   A(int n=1) : _n(n) {};
   void setVal(int val = 1);
   void setVal(std::string, int val = 1)
private : 
   int _n;
};

I can't see how to declare a default argument in an overloaded method and also in the constructor. 
I tried using "n"_a=1 and arg("n")=0 but none work, g++ crashes with a lot of errors. 

I'd really appreciate any help.

**Comment by jiwaszki** [comment]
Hi @hbadi , first of a ctor should be public. Are you sure that you tried it according to the docs? Here you have a working sample:
```cpp
#include <string>

#include <pybind11/pybind11.h>

namespace py = pybind11;

class A
{
public:
  A(int n = 1) : _n(n){};
  void setVal(int val = 1);
  void setVal(std::string, int val = 1);

private:
  int _n;
};

PYBIND11_MODULE(foo, module)
{
  py::class_<A, std::shared_ptr<A> > cls(module, "A");
  cls.def(py::init<int>(), py::arg("n") = 0);
  // other bindings...
}
```

**Comment by hbadi** [comment]
Hi @jiwaszki, thanks for your answer
It really helps me, i forgot to list all arguments cause in my case the constructor have two arguments and only one with a default value !
Is it possible to only list the arguments which have a default value like ? : 
`cls.def(py::init<int,int>(), py::arg("n") = 0);`

Thanks.

**Reply by jiwaszki** [reply]
No, it will throw with `"The number of argument annotations does not match the number of function arguments"` -- if you are using argument annotations it must match number of arguments in original function.
